Add PVC storage to LimitRange

markturansky opened the pull request:
This PR adds the ability to add a LimitRange to a namespace that enforces min/max on `pvc.Spec.Resources.Requests["storage"]`.

@derekwaynecarr @abhgupta @kubernetes/sig-storage 

Examples forthcoming. 

``` release-note
pvc.Spec.Resources.Requests min and max can be enforced with a LimitRange of type "PersistentVolumeClaim" in the namespace
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30145)

<!-- Reviewable:end -->

derekwaynecarr commented:
Have not finished a deep review, but this will also require doc updates somewhere around these two areas:

https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/admin/resourcequota
https://github.com/kubernetes/kubernetes.github.io/blob/master/docs/admin/limitrange

derekwaynecarr commented on plugin/pkg/admission/limitranger/admission_test.go:
```diff
@@ -30,14 +30,17 @@ import (
 )
 
-func getResourceList(cpu, memory string) api.ResourceList {
+func getResourceList(cpu, memory, storage string) api.ResourceList {
```
this is a style thing, but since we will not call with all 3 values populated, can we have something like:

```
getComputeResourceList(cpu,memory string)
getStorageResourceList(storage string)
```

It will help us keep things sensible if storage ever does more, and wont confuse compute resource list.

markturansky replied:
I originally wanted to do something like that but thought I was being too lazy, so I edited every single func call ...

derekwaynecarr replied:
we follow the same pattern in validation_test.go, i just think its hard to read 3 strings ;-)

markturansky replied:
Funcs split as suggested.

derekwaynecarr commented on pkg/api/v1/types.go:
```diff
@@ -2966,6 +2966,8 @@ const (
 	// Limit that applies to all containers in a namespace
 	LimitTypeContainer LimitType = "Container"
+	// Limit that applies to all persistent volume claims in a namespace
+	LimitTypePersistentVolumeClaim LimitType = "PersistentVolumeClaim"
```
can you add something to validation_test.go that tests with this limit type?

derekwaynecarr commented:
can you also add a commit that updates: https://github.com/kubernetes/kubernetes/blob/master/docs/design/admission_control_limit_range.md

markturansky committed:
add pvc storage to LimitRange

markturansky commented on plugin/pkg/admission/limitranger/admission_test.go:
```diff
@@ -634,3 +642,64 @@ func TestLimitRangerCacheAndLRUExpiredMisses(t *testing.T) {
+	successCases := []testCase{
+		{
+			pvc:        validPersistentVolumeClaim("pvc-is-min-storage-request", getResourceRequirements(getStorageResourceList("1Gi"), getStorageResourceList(""))),
+			limitRange: createLimitRange(api.LimitTypePersistentVolumeClaim, getStorageResourceList("1Gi"), api.ResourceList{}, api.ResourceList{}, api.ResourceList{}, api.ResourceList{}),
```
@derekwaynecarr regarding your comment about using `api.LimitTypePersistentVolumeClaim` in  a test, doesn't this do it? The `LimitRange` is created w/ this type and then `PersistentVolumeClaimLimitFunc` looks specifically for this type.

derekwaynecarr replied:
i meant a validation_test.

btw, i suspect our QA team will potentially open bz reports if we do not validate against setting fields not yet supported by the admission plugin.  for example, they may question if we should error when a user sets a default or defaultRequest value for a pvc in limit range, and want a validation error.  i am not sure how concerned i am to check for that.

whats your opinion?

markturansky replied:
What is the established pattern for unsupported fields?  Silently ignored (docs discuss them) or fail?

I'd like to go with whatever standard is in place. If no standard, I am happy with silent ignoring. I suppose one could argue failure is explicit while silently ignoring is implicit and therefore not obvious.

derekwaynecarr replied:
We ignore for our downstream uses today that limit our OpenShift resources... I am fine continuing that now for here as well.

markturansky commented:
@derekwaynecarr the LimitRange doc was updated per your request and I also added an example showing the use of both LimitRange and StorageQuota within a namespace.  I tested them together locally and they work well.

markturansky commented:
@derekwaynecarr anything left to do on this PR?

derekwaynecarr commented on docs/design/admission_control_limit_range.md:
```diff
@@ -209,6 +210,25 @@ Across all containers in pod, the following must hold true
+
+Supported Constraints:
+
+Across all claims in a namespace, the follow must hold true:
```
nit: s/follow/following

markturansky replied:
nit addressed

derekwaynecarr commented on docs/design/admission_control_limit_range.md:
```diff
@@ -209,6 +210,25 @@ Across all containers in pod, the following must hold true
+| Min | Min >= Request (required) |
+| Max | Max <= Request (required) |
+
+Supported Defaults: None. Storage is a required in `PersistentVolumeClaim`, so defaults are not applied at this time.
```
I am surprised by this.

markturansky replied:
Capacity is a required field on PVC, so what value would be defaulted in?

markturansky replied:
Let me correct that sentence, too...

derekwaynecarr commented on examples/volumes/limits_and_quota/README.md:
```diff
@@ -0,0 +1,53 @@
+# Limiting storage resource consumption
```
doesn't this type of document (which is good) belong in the kubernetes.github.io repo instead?

markturansky replied:
both?  I think I need to edit the docs in that repo for LimitRange as I did the design doc in this repo.

derekwaynecarr commented on docs/design/admission_control_limit_range.md:
```diff
@@ -209,6 +210,25 @@ Across all containers in pod, the following must hold true
+| Min | Min >= Request (required) |
+| Max | Max <= Request (required) |
+
+Supported Defaults: None. Storage is a required field in `PersistentVolumeClaim`, so defaults are not applied at this time.
```
Many fields that are required are assigned server side if omitted, I had thought you needed that use case.  

It's fine not doing it now, I was just surprised.

markturansky replied:
Oh, I see. An admin might assign 1Gi as the default and that is applied *before* validation.

I wasn't specifically asked for that but I can add it if you think it is important to follow an established pattern.

derekwaynecarr replied:
defaulting support can come later when or if there is an actual need.

derekwaynecarr commented:
couple of questions, request to move the detailed doc to a separate repo, and your input on validation of limit range using this type.

markturansky commented:
@derekwaynecarr added validation for this limitRangeType. Removed the detailed doc for a move to the docs repo.

I don't have a requirement for a default storage size, but can add if necessary. I tend to like the user getting told what the min/max is, as opposed to getting a default value. But I'll gladly do whatever the general UX of Kube is for this.

Squashed for good measure and will follow up with the docs repo PR.

markturansky commented:
@derekwaynecarr I added more to validation_test.go and everything looks good to go with this LimitRange. Please let me know if I'm missing anything else.

markturansky commented:
Docs adds https://github.com/kubernetes/kubernetes.github.io/pull/1089

markturansky commented:
@derekwaynecarr is this PR still on track for 1.4? We're relying on it downstream.  Please let me know if there's anything else. Otherwise, LGTM?

eparis commented:
ok to test

derekwaynecarr commented:
@markturansky - we are past feature freeze for kube 1.4.  I am fine with this PR for 1.5 as-is.  I am copying @kubernetes/api-review-team for their ack in 1.5 to ack on LimitRange enforcing PersistentVolumeClaim constraints.

markturansky commented:
@derekwaynecarr Any further work required for this? Docs got merged and I rebased this PR for good measure.

derekwaynecarr commented:
Don't think so, will review shortly.

On Thursday, September 22, 2016, Mark Turansky <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> Any further work
> required for this? Docs got merged and I rebased this PR for good measure.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30145#issuecomment-248969000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF8dbPglddUrb45Px2TR_LnSfNB4cDJVks5qsrjogaJpZM4Jd0E6>
> .
>

derekwaynecarr commented:
LGTM

derekwaynecarr commented:
if/when we add support to break `cpu` and `memory` into `requests.cpu` and `limits.cpu` so we can support a discrete min/max range on the request separate from the limit, we can do the same for `storage`.

markturansky commented:
@derekwaynecarr I looked through the tests and ran them locally. I'm not sure if these are flakes or not, but the test failures seem unrelated to the changes in this PR.

I rebased for good measure.

markturansky commented:
@derekwaynecarr tests are all green now. i think it was just flakes before. can you re-add LGTM after my rebase?

markturansky commented:
@eparis can you check this PR out? It seems to be missing my CLA? Also can use the LGTM re-added. I lost it after a rebase.
